### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by mb-dbc-dk**
Annotation and example application

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+...
+    }
+
+```
+
+The calls are across bean-boundries, so interceptors are in play. The environment variable `BACKEND_TIMING` is set to some duration (ex. `20ms`), it is expected that `getDatabaseEntriesFor()` takes under normal circumstances 60% of the wall clock time, and `processEntries()` about 40%.
+This will log all the arguments and exceptions but not return values for the calls to `getDatabaseEntriesFor()`, that take more than 12ms, as an error.
+And for `processEntries()` calls that take more that 8ms it will log trackingId and return value.
+
+This is bould upon the `slf4j` log framework. And the logger used is called `SlowCallLog`. An mdc value called: `call_duration_ms` records duration in milliseconds.
```
Comment (mbd-dbc-dk) [inline README.md]: build

File: README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+...
+        List<Row> rows = backend.getDatabaseEntriesFor(key, whenish);
+...
+        backend.processEntries(rows, trackingId);
+...
+    }
+
+```
+
+The calls are across bean-boundries, so interceptors are in play. The environment variable `BACKEND_TIMING` is set to some duration (ex. `20ms`), it is expected that `getDatabaseEntriesFor()` takes under normal circumstances 60% of the wall clock time, and `processEntries()` about 40%.
```
Comment (noah-dbc) [inline README.md]: boundaries

File: src/main/java/dk/dbc/commons/slowcalllog/NanoUnit.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (mbd-dbc-dk) [inline src/main/java/dk/dbc/commons/slowcalllog/NanoUnit.java:1]: Det virker som lidt overkill at have alle disse units til dette formål. Hvorfor ikke bare angive tallet i f.eks. fractionelle ms? I praksis dækker det vel alt hvad man vll have brug for?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+Ideally you would have debug log level for the calls that are slow, but not those that are fast. This cannot be achieved since you only know after the method call if you wanted one or the other. Theoretically you can make a logger that accumulates all the lines, that are of debug level, and posts them when the call is determined to be slow, however this approach will take up a log of resources both memory and cpu cycles, that potentially will render all the calls slow.
+
+Instead we opt to log when a call has been completed, and determined slow.
+
+__WARNING__: We log the arguments as they are when the call is completed. If arguments are altered by the method, they will not be seen as the input edition of them. This might hide the culprit for the call duration.
+
+It is all based around an annotation `@SlowCallLog`, and `CDI` in `EE`. During bean discovery in the container, all `@SlowCallLog` annotations are picked up. An interceptor is configured with a wrapper for the given methods, that calls the method, and if the calls duration is above a threshold, the call in logged with it's parameters and optionally `return value` or `exception` if the call fails.
```
Comment (noah-dbc) [inline README.md]: no apostrophe in "it's"

File: example/README.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
+# Example Payara Application
+
+An example that uses `dbc-commons-slow-call-log`
+
+
+## Building
+
+Remember to build library first (and install it) by running: `(cd .. && mvn install)`
+
+The you can build the JakartaEE application by ruinning: `mvn package` - this also downloads a `payara-micro.jar`
```
Comment (noah-dbc) [inline example/README.md]: running

File: src/main/java/dk/dbc/commons/slowcalllog/SlowCallLog.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+     * Scale of value from {@link #env()}
+     * <p>
+     * For splitting time
+     * <p>
+     * This is useful if you set the value in {@link #env()} to 2ms which is all
+     * the time a request is allowed to take. Then you annotate the GET/POST
+     * method with 'scale = 1.0', if it has 3 sub calls you can annotate them
+     * with scale proportional to the time they should take under normal
+     * circumstances. Then you'll get logging for each of the methods that takes
+     * a disproportional amount of time.
```
Comment (noah-dbc) [inline src/main/java/dk/dbc/commons/slowcalllog/SlowCallLog.java]: "disproportionate"

File: src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogExtension.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2020 DBC A/S (http://dbc.dk/)
+ *
+ * This is part of slow-call-log [unloadable]
```
Comment (noah-dbc) [inline src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogExtension.java]: what's with "unloadable" here?

File: src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogInterceptor.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+                if (duration >= thresholdInNs)
+                    exception.call(duration, ic.getParameters(), makeExceptionString(ex));
+                throw ex;
+            }
+        };
+    }
+
+    /**
+     * If en exception doesn't have a message find the one that does, and give a
+     * meaning full message
```
Comment (noah-dbc) [inline src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogInterceptor.java]: meaningful

File: src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogInterceptor.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    static long durationInNs(String duration) {
+        try {
+            String[] parts = duration.split("(?=\\D)", 2);
+            if (parts.length != 2)
+                throw new IllegalArgumentException("Don't know duration: " + duration);
+            long amount = Long.parseUnsignedLong(parts[0].trim());
+            return amount * NanoUnit.of(parts[1].trim()).nanoSeconds();
+        } catch (IllegalArgumentException ex) {
+            throw new IllegalArgumentException("Don't know duration: " + duration);
+
```
Comment (noah-dbc) [inline src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogInterceptor.java]: no need for empty line here

**Review by noah-dbc — CHANGES_REQUESTED** [review]
typos, mostly.

File: src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogInterceptor.java [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+                throw new IllegalArgumentException("Cannot have positional parameters greater than or equal to " + method.getParameterCount());
+        }
+        return params;
+    }
+
+    /**
+     * Construct an invoker that logs if duration is too long
+     *
+     * @param thresholdInNs how many nanoseconds to allow call to take
+     * @param logger        how to log is duration is exceeded, and call
```
Comment (mbd-dbc-dk) [inline src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogInterceptor.java]: if duration

File: src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogInterceptor.java [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+     *                      succeeded
+     * @param exception     how to log is duration is exceeded, and call failed
```
Comment (mbd-dbc-dk) [inline src/main/java/dk/dbc/commons/slowcalllog/SlowCallLogInterceptor.java]: if duration

File: README.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+A fully running payara-micro demonstration can be found in the [example](example) directory.
+
+It is based around this structure:
+
+```
+@Stateless // or any other bean annotation
+public MyBackendClass {
+
+    @SlowCallLog(env="BACKEND_TIMING", factor=.6, result=false, level="ERROR")
```
Comment (mbd-dbc-dk) [inline README.md]: factor .. det ser ud som om at det endte med at hedde scale i den endelige udgave?

File: src/main/java/dk/dbc/commons/slowcalllog/SlowCallLog.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+     * default is {-1} which is everything. use {} to not log any call
+     * parameters
+     *
+     * @return {-1}
+     */
+    @Nonbinding
+    int[] parameters() default {-1};
+
+    /**
+     * Should log return value, if not void, too?
```
Comment (mbd-dbc-dk) [inline src/main/java/dk/dbc/commons/slowcalllog/SlowCallLog.java]: "Set to true, if return value should be logged. This settings is ignored if the method has void as return type."

**Review by mbd-dbc-dk — CHANGES_REQUESTED** [review]

**Review by thpdbc — APPROVED** [review]

**Review by noah-dbc — APPROVED** [review]

**Pull request merged**